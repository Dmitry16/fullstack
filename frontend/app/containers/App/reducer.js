/*
 * AppReducer
 *
 * The reducer takes care of our data. Using actions, we can change our
 * application state.
 * To add a new action, add it to the switch statement in the reducer function
 *
 * Example:
 * case YOUR_ACTION_CONSTANT:
 *   return state.set('yourStateVariable', true);
 */

import { fromJS } from 'immutable';

import {
  LOAD_MOVIES_SUCCESS,
  LOAD_MOVIES,
  LOAD_MOVIES_ERROR,
} from './constants';

// The initial state of the App
const initialState = fromJS({
  loading: false,
  error: false,
  currentMovieTitle: false,
  movieData: {
    movies: false,
  },
});

function appReducer(state = initialState, action) {
  switch (action.type) {
    case LOAD_MOVIES:
      return state
        .set('loading', true)
        .set('error', false)
        .setIn(['movieData', 'movies'], false)
    case LOAD_MOVIES_SUCCESS:
      return state
        .setIn(['movieData', 'movies'], action.movies)
        .set('loading', false)
        .set('currentMovieTitle', action.currentMovieTitle);
    case LOAD_MOVIES_ERROR:
      return state
        .set('error', action.error)
        .set('loading', false);
    default:
      return state;
  }
}

export default appReducer;
