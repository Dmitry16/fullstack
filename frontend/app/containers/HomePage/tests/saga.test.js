/**
 * Tests for HomePage sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { LOAD_MOVIES } from 'containers/App/constants';
import { moviesLoaded, movieLoadingError } from 'containers/App/actions';

import moviesData, { getMovies } from '../saga';

const movieTitle = 'Blade Runner';

/* eslint-disable redux-saga/yield-effects */
describe('getMovies Saga', () => {
  let getMoviesGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getMoviesGenerator = getMovies();

    const selectDescriptor = getMoviesGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getMoviesGenerator.next(movieTitle).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  // it('should dispatch the moviesLoaded action if it requests the data successfully', () => {
  //   const response = [{
  //     title: 'First movie',
  //   }, {
  //     title: 'Second movie',
  //   }];
  //   const putDescriptor = getMoviesGenerator.next(response).value;
  //   expect(putDescriptor).toEqual(put(moviesLoaded(response, movieTitle)));
  // });

  it('should call the movieLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getMoviesGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(movieLoadingError(response)));
  });
});

describe('moviesData Saga', () => {
  const moviesDataSaga = moviesData();

  it('should start task to watch for LOAD_REPOS action', () => {
    const takeLatestDescriptor = moviesDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_MOVIES, getMovies));
  });
});
