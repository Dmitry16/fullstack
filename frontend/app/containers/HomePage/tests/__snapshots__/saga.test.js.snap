// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getMovies Saga should call the movieLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getMovies Saga should call the movieLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://localhost:3003/api/search?keyword=Blade Runner",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
